parlai/core/torch_agent.py

```diff
@@ -814,11 +833,15 @@ def init_optim(self, params, optim_states=None, saved_optim_type=None):
                 'switch to --optimizer adam.'
             )
 
+        # TODO: if self.fp16 and self.fp16_impl == 'mem_efficient'... check that class is ok
         optim_class = self.optim_opts()[opt['optimizer']]
         self.optimizer = optim_class(params, **kwargs)
         if self.fp16:
-            self.optimizer = fp16_optimizer_wrapper(self.optimizer)
-
+            if self.fp16_impl == 'apex':
+                self.optimizer = fp16_optimizer_wrapper(self.optimizer)
+            else:
```

Comment:
nit: `elif self.fp16_impl == 'mem_efficient'` here in case we want to add further choices down the line?